gov/politeia/proposals.go

```diff
@@ -91,373 +90,607 @@ func NewProposalsDB(politeiaURL, dbPath string) (*ProposalDB, error) {
+func (db *ProposalsDB) fetchTicketVoteSummaries(tokens []string) (map[string]ticketvotev1.Summary, error) {
+	voteSummaries := make(map[string]ticketvotev1.Summary, len(tokens))
+	paginatedTokens := paginateTokens(tokens, ticketvotev1.SummariesPageSize)
+	for i := 0; i < len(paginatedTokens); i++ {
+		sr, err := db.client.TicketVoteSummaries(ticketvotev1.Summaries{
+			Tokens: paginatedTokens[i],
+		})
+		if err != nil {
+			return nil, fmt.Errorf("Pi client TicketVoteSummaries err: %w", err)
+		}
+		for token := range sr.Summaries {
+			voteSummaries[token] = sr.Summaries[token]
+		}
 	}
+	return voteSummaries, nil
+}
 
-	var queryParam string
-	if len(lastProposal) > 0 && lastProposal[0].TokenVal != "" {
-		queryParam = fmt.Sprintf("?before=%s", lastProposal[0].TokenVal)
+// paginateTokens paginates tokens in a matrix according to the provided
+// page size.
+func paginateTokens(tokens []string, pageSize uint32) [][]string {
+	n := len(tokens)/int(pageSize) + 1 // number of pages needed
+	ts := make([][]string, n)
+	page := 0
+	for i := 0; i < len(tokens); i++ {
+		if len(ts[page]) >= int(pageSize) {
+			page++
+		}
+		ts[page] = append(ts[page], tokens[i])
 	}
-	publicProposals, err := db.fetchAPIData(queryParam)
+	return ts
+}
+
+// fetchTicketVoteResults fetches the vote data for the given proposal token,
+// then builds and returns its parsed chart data.
+func (db *ProposalsDB) fetchTicketVoteResults(token string) (*pitypes.ProposalChartData, error) {
+	// Fetch ticket votes details to acquire vote bits options info.
+	details, err := db.client.TicketVoteDetails(ticketvotev1.Details{
+		Token: token,
+	})
 	if err != nil {
-		return err
+		return nil, fmt.Errorf("Pi client TicketVoteDetails err: %w", err)
+	}
+
+	// Maps the vote bits option to their respective string ID.
+	voteOptsMap := make(map[uint64]string)
+	for _, opt := range details.Vote.Params.Options {
+		voteOptsMap[opt.Bit] = opt.ID
 	}
 
-	n, err := db.saveProposals(publicProposals)
+	tvr, err := db.client.TicketVoteResults(ticketvotev1.Results{
+		Token: token,
+	})
 	if err != nil {
-		return err
+		return nil, fmt.Errorf("Pi client TicketVoteResults err: %w", err)
+	}
+
+	// Parse proposal chart data from the ticket vote results reply and
+	// sort it afterwords.
+	type voteData struct {
+		yes, no   uint64
+		timestamp int64
+	}
+	votes := make([]*voteData, 0, len(tvr.Votes))
+	for iv := range tvr.Votes {
+		// Vote bit comes as a hexadecimal number in the format of a string.
+		// Convert it to uint64.
+		bit, err := strconv.ParseUint(tvr.Votes[iv].VoteBit, 16, 64)
+		if err != nil {
+			return nil, err
+		}
+
+		// Verify vote bit is valid.
+		err = voteBitVerify(details.Vote.Params.Options,
+			details.Vote.Params.Mask, bit)
+		if err != nil {
+			return nil, err
+		}
+
+		// Parse relevant data.
+		var vd voteData
+		switch voteOptsMap[bit] {
+		case "yes":
+			vd.yes = 1
+			vd.no = 0
+		case "no":
```

Comment:
Pls use `ticketvotev1.VoteOptionIDApprove` and `ticketvotev1.VoteOptionIDReject` in these cases